keylime/src/structures/capabilities_negotiation.rs

```diff
@@ -0,0 +1,353 @@
+use serde::{Deserialize, Serialize};
+
+// Define the structure for the AttestationRequest:
+#[derive(Serialize, Deserialize, Debug)]
+pub struct AttestationRequest {
+    #[serde(rename(serialize = "data", deserialize = "data"))]
+    pub data: RequestData,
+}
+#[derive(Serialize, Deserialize, Debug)]
+pub struct RequestData {
+    #[serde(rename(serialize = "type", deserialize = "type"))]
+    pub type_: String,
+    pub attributes: Attributes,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Attributes {
+    pub evidence_supported: Vec<EvidenceSupported>,
+    pub boot_time: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct EvidenceSupported {
+    pub evidence_class: String,
+    pub evidence_type: String,
+    pub agent_capabilities: AgentCapabilities,
+    pub version: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct AgentCapabilities {
+    pub spec_version: String,
+    pub hash_algorithms: Vec<String>,
+    pub signing_schemes: Vec<String>,
+    pub attestation_keys: Vec<AttestationKeys>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct AttestationKeys {
+    pub key_class: String,
+    pub key_identifier: String,
+    pub key_algorithm: String,
+    pub public_hash: String,
+}
+
+// Define the structure for the AttestationResponse:
+#[derive(Serialize, Deserialize, Debug)]
+pub struct AttestationResponse {
+    #[serde(rename(serialize = "data", deserialize = "data"))]
+    pub data: ResponseData,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ResponseData {
+    #[serde(rename(serialize = "type", deserialize = "type"))]
+    pub type_: String,
+    pub attributes: ResponseAttributes,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ResponseAttributes {
+    pub evidence_collected: Vec<EvidenceCollected>,
+    pub boot_time: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct EvidenceCollected {
+    pub evidence_class: String,
+    pub evidence_type: String,
+    pub chosen_parameters: ChosenParameters,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ChosenParameters {
+    pub nonce: String,
+    pub pcr_selection: Vec<i32>,
+    pub hash_algorithm: String,
+    pub signing_scheme: String,
+    pub attestation_key: AttestationKey,
+    pub starting_offset: Option<i32>,
+}
```

Comment:
This structure is mixing parameters from different types of evidence: `starting_offset` is related with the IMA log (in the example it is called a `partial_log`), while the rest of the fields are related with the TPM (and TPM quote).

Maybe it should be an union of the different types of parameters that are related with each type of evidence supported.

Reply:
The problem with unions is how they are "selected" without messing the JSON ... you are completely right that some of the fields only apply to TPM, and starting_offset only applies to IMA but ... why not letting all them as an Option and let serde code them as required?

Reply:
To improve code readability and avoid confusion. Also we want to take advantage of using the structures as a form of validation.

If you disagree, let's discuss in the meeting

Reply:
Initially, I implemented the solution by using Option<>. I was not able to find a solution by using enums.